creator-node/src/middlewares/premiumContent/premiumContentMiddleware.ts

```diff
@@ -31,70 +28,71 @@ export const premiumContentMiddleware = async (
   next: NextFunction
 ) => {
   try {
-    const {
-      signedDataFromDiscoveryNode,
-      signatureFromDiscoveryNode,
-      signedDataFromUser,
-      signatureFromUser
-    } = req.headers
-
-    if (
-      !signedDataFromDiscoveryNode ||
-      !signatureFromDiscoveryNode ||
-      !signedDataFromUser ||
-      !signatureFromUser
-    ) {
+    const cid = req.params && req.params.CID
+    if (!cid) {
       return sendResponse(
         req,
         res,
-        errorResponseBadRequest('Missing request headers.')
+        errorResponseBadRequest(`Invalid request, no CID provided.`)
       )
     }
 
+    const premiumContentHeaders = req.headers['x-premium-content'] as string
+    const premiumContentAccessChecker = req.app.get(
+      'premiumContentAccessChecker'
+    ) as PremiumContentAccessChecker
+    const libs = req.app.get('audiusLibs')
+    const redis = req.app.get('redisClient')
     const logger = (req as any).logger as Logger
 
-    const discoveryNodeWallet = recoverWallet(
-      signedDataFromDiscoveryNode,
-      signatureFromDiscoveryNode
-    )
-    const isRegisteredDN = await isRegisteredDiscoveryNode({
-      wallet: discoveryNodeWallet,
-      libs: req.app.get('audiusLibs'),
-      logger,
-      redis: req.app.get('redisClient')
-    })
-    if (!isRegisteredDN) {
-      return sendResponse(
-        req,
-        res,
-        errorResponseForbidden(
-          'Failed discovery node signature validation for premium content.'
-        )
-      )
+    const { doesUserHaveAccess, trackId, isPremium, error } =
+      await premiumContentAccessChecker.checkPremiumContentAccess({
+        cid,
+        premiumContentHeaders,
+        libs,
+        logger,
+        redis
+      })
+    if (doesUserHaveAccess) {
+      // Set premium content track id and 'premium-ness' so that next middleware or
+      // request handler does not need to make trips to the database to get this info.
+      // We need the info because if the content is premium, then we need to set
+      // the cache-control response header to no-cache so that nginx does not cache it.
+      ;(req as any).premiumContent = {
+        trackId,
+        isPremium
+      }
+      next()
+      return
+    } else {
+      switch (error) {
+        case PremiumContentAccessError.MISSING_HEADERS:
+          return sendResponse(
+            req,
+            res,
+            errorResponseForbidden(
+              'Missing request headers for premium content.'
```

Comment:
Is it a good idea to expose these error messages to the client?

Reply:
i dont think it really hurts, it's not going to help potential bad users much but can help us debug. lmk if you feel strongly

Reply:
My take is the server should reveal as little info as possible and be opaque as possible about what failed. I'm not a huge proponent, but I've even heard suggestions of using 404s in scenarios like this, to not let the user know that the resource they don't have access to exists.

I'm open to leaving it for now for ease of debugging while building this out, but long term I think we should log the exact failure and give an opaque 403 to the client.

(Arguably this first one should be a 401 Unauthorized, but that's slightly pedantic 😄 )

Reply:
fair fair